### PR TITLE
Add BART Backbone

### PR DESCRIPTION
**Pull request by abheesht17**
Resolves https://github.com/keras-team/keras-nlp/issues/649

Conversion Notebook: https://colab.research.google.com/drive/1JeWNXHFEdvFpxV_G2PDy5AAk6R6eLdJz?usp=sharing

### CODE REVIEW

File: keras_nlp/layers/transformer_decoder.py [hunk 1]
```diff
@@ -165,7 +165,7 @@ def _build(self, input_shape, has_cross_attention):
             self._cross_attention_layer = keras.layers.MultiHeadAttention(
                 num_heads=self.num_heads,
                 key_dim=head_dim,
-                value_dim=hidden_dim,
+                value_dim=head_dim,
```
Comment (jbischof) [inline keras_nlp/layers/transformer_decoder.py:168]: Whoa, did you find a bug here? If so, let's discuss and break this into a different PR.

File: keras_nlp/models/bart/bart_backbone.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+        "segment_ids": tf.constant(
+            [0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0], shape=(1, 12)
+        ),
+        "padding_mask": tf.constant(
+            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0], shape=(1, 12)
+        ),
+    }
+
+    # Pretrained BART encoder-decoder model
+    model = keras_nlp.models.BartBackbone.from_preset("base_base_en_uncased")
```
Comment (jbischof) [inline keras_nlp/models/bart/bart_backbone.py:75]: This preset name can't be right 🤡

File: keras_nlp/models/bart/bart_backbone.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+        # Token embedding layer. This layer is shared by encoder and decoder.
+        token_embedding_layer = keras.layers.Embedding(
+            input_dim=vocabulary_size,
+            output_dim=hidden_dim,
+            embeddings_initializer=bart_kernel_initializer(),
+            name="token_embedding",
+        )
+
+        # Encoder
```
Comment (jbischof) [inline keras_nlp/models/bart/bart_backbone.py:127]: Could we make these section headers more visually distinct from other comments? One idea:

```python
##### Encoder #####
```

**Review by jbischof — COMMENTED** [review]
Looking great, some quick comments. What is going on with our `TransformerDecoder`!?

File: keras_nlp/models/bart/bart_backbone.py (hunk 2 above)
Comment (abheesht17) [reply keras_nlp/models/bart/bart_backbone.py:75]: Oh, yeah. This mistake is actually present in `BertBackbone`: https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/models/bert/bert_backbone.py#L80. So, this mistake persisted since I copied over basic stuff from there, lol. I have fixed all errors of this kind in this PR: https://github.com/keras-team/keras-nlp/pull/640.

File: keras_nlp/models/bart/bart_backbone.py (hunk 3 above)
Comment (abheesht17) [reply keras_nlp/models/bart/bart_backbone.py:127]: The issue is that this doesn't work with flake8 :P

```
E266 too many leading '#' for block comment
```

Comment (jbischof) [reply keras_nlp/models/bart/bart_backbone.py:125]: OK, don't upset our robot overloads 🤖 !

File: keras_nlp/models/bart/bart_backbone.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+from keras_nlp.layers.transformer_encoder import TransformerEncoder
+from keras_nlp.models.backbone import Backbone
+
+
+def bart_kernel_initializer(stddev=0.02):
+    return keras.initializers.TruncatedNormal(stddev=stddev)
+
+
+@keras.utils.register_keras_serializable(package="keras_nlp")
+class BartBackbone(Backbone):
```
Comment (mattdangerw) [inline keras_nlp/models/bart/bart_backbone.py:31]: I wish they had chosen a name that wasn't so visually similar to Bert. This is going to confuse me so much :P

File: keras_nlp/models/bart/bart_backbone.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        # Decoder inputs.
+        decoder_token_id_input = keras.Input(
+            shape=(None,), dtype="int32", name="decoder_token_ids"
+        )
+        decoder_padding_mask = keras.Input(
+            shape=(None,), dtype="int32", name="decoder_padding_mask"
+        )
+
+        # Token embedding layer. This layer is shared by encoder and decoder.
+        token_embedding_layer = keras.layers.Embedding(
```
Comment (mattdangerw) [inline keras_nlp/models/bart/bart_backbone.py:118]: When BART is used as a language model, are the embedding weights shared for the output projection? Or is there a separate set of parameters used?

File: keras_nlp/models/bart/bart_backbone.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
+# Copyright 2022 The KerasNLP Authors
```
Comment (mattdangerw) [inline keras_nlp/models/bart/bart_backbone.py:1]: Let's add tests!

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Just some quick initial comments!

File: keras_nlp/models/bart/bart_backbone.py (hunk 4 above)
Comment (abheesht17) [reply keras_nlp/models/bart/bart_backbone.py:31]: :joy:

File: keras_nlp/models/bart/bart_backbone.py (hunk 5 above)
Comment (abheesht17) [reply keras_nlp/models/bart/bart_backbone.py:118]: @mattdangerw - they do define separate Embedding layers in both the encoder and decoder: https://github.com/huggingface/transformers/blob/44caf4f6f47120a4aaca561c9cc8041455bef705/src/transformers/models/bart/modeling_bart.py#L723-L726 and https://github.com/huggingface/transformers/blob/44caf4f6f47120a4aaca561c9cc8041455bef705/src/transformers/models/bart/modeling_bart.py#L896-L899.

But they copy over the weights for these layers from this layer: https://github.com/huggingface/transformers/blob/main/src/transformers/models/bart/modeling_bart.py#L1169. 

Moreover, when I printed the trainable layers, these two embedding layers did not show up: https://colab.research.google.com/drive/15871_w5LqU9-wzykFbb8dAne6g4zvyat?usp=sharing. So, I think the same set of parameters are being used?

File: keras_nlp/models/bart/bart_backbone.py (hunk 6 above)
Comment (abheesht17) [reply keras_nlp/models/bart/bart_backbone.py:1]: Will do!

**Review by abheesht17 — COMMENTED** [review]
@mattdangerw, replied to the comment about embedding layers.

File: keras_nlp/models/bart/bart_backbone.py (hunk 4 above)
Comment (jbischof) [reply keras_nlp/models/bart/bart_backbone.py:31]: I wish they explained what BART stands for....

image

Comment (abheesht17) [reply keras_nlp/models/bart/bart_backbone.py:31]: BERT -> Bidirectional Encoder Representations from Transformers
BART -> Bidirectional **Auto-encoder** Representations from Transformers?

Guessing...:P

Comment (jbischof) [reply keras_nlp/models/bart/bart_backbone.py:31]: Interesting! Couldn't find it in the paper

**Review by jbischof — APPROVED** [review]
Just need to rebase on #667 and we should be able to merge

File: keras_nlp/layers/transformer_decoder.py (hunk 1 above)
Comment (abheesht17) [reply keras_nlp/layers/transformer_decoder.py]: Done!

File: keras_nlp/models/bart/bart_backbone.py (hunk 5 above)
Comment (mattdangerw) [reply keras_nlp/models/bart/bart_backbone.py:117]: Great! Great, this is the easy case for us, where the backbone is self sufficient for language modeling tasks. We have prior art for this with BERT pretraining and Chen's upcoming GPT2 language model.

File: keras_nlp/models/bart/bart_backbone.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+
+
+@keras.utils.register_keras_serializable(package="keras_nlp")
+class BartBackbone(Backbone):
+    """BART encoder-decoder network.
+
+    This class implements a Transformer-based encoder-decoder model as
+    described in
+    ["BART: Denoising Sequence-to-Sequence Pre-training for Natural Language Generation, Translation, and Comprehension"](https://arxiv.org/abs/1910.13461).
+    It includes the embedding lookups and transformer layers.
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply keras_nlp/models/bart/bart_backbone.py]: Oh, yeah. Brain fart, indeed 🤦🏼‍♂️

File: keras_nlp/models/bart/bart_backbone.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+            output_dim=hidden_dim,
+            embeddings_initializer=bart_kernel_initializer(),
+            name="token_embedding",
+        )
+
+        # Encoder
+
+        # Embed tokens and positions.
+        token_embedding = token_embedding_layer(encoder_token_id_input)
+        position_embedding = PositionEmbedding(
```
Comment (mattdangerw) [inline keras_nlp/models/bart/bart_backbone.py:128]: Let's add a comment explaining that the position embedding is not shared, but the token embedding is.

**Review by mattdangerw — APPROVED** [review]
Just some minor comments. I can fix as I merge.
